### PR TITLE
Rework jkind default

### PR DESCRIPTION
**Pull request by alanechang**

### CODE REVIEW

**Comment by alanechang** [comment]
This follows #2107 and reworks how type variables get their jkinds.

Prior to #2107, type variables start with jkind `any` and `->` (`Tarrow`) forces argument and result types to concrete jkinds (represented as sort vars).

#2107 passes the tests by changing type variables to all start with concrete jkinds.

This PR makes the type variable jkind initialization controllable through a new field in `TyVarEnv.policy` (either `Sort` or `Any`) and uses different policies depending on the typing context.

The general rule it follows is that

- calls from `typecore` initialize with `Any`
- calls from `typedecl` _mostly_ initialize with `Sort`. `Any` is used for type manifest where the variables get unified with type parameters of concrete jkinds in the end and constraint clauses (this break backward compatibility, see `test_flat.ml`).
- calls from `typeclass` initialize with `Sort`

This PR also contains a change to `Typetexp.transl_type_var` that makes jkind annotation on type variables behave as upper bounds.

## How to review

Might be best to first review the changes in `typetexp` and then look at the various call sites of `transl_simple_type`.

File: ocaml/typing/typedecl.ml [hunk 1]
```diff
@@ -468,7 +468,7 @@ let make_constructor
             transl_constructor_arguments env univars closed sargs
           in
           let tret_type =
-            transl_simple_type env ?univars ~closed Mode.Alloc.Const.legacy
+            transl_simple_type ~new_var_jkind:Sort env ?univars ~closed Mode.Alloc.Const.legacy
```
Comment (goldfirere) [inline ocaml/typing/typedecl.ml:471]: `Sort` is good here, as we're always in a GADT.

File: ocaml/typing/typedecl.ml [hunk 2]
```diff
@@ -626,8 +626,8 @@ let transl_declaration env sdecl (id, uid) =
   let params = List.map (fun (cty, _) -> cty.ctyp_type) tparams in
   let cstrs = List.map
     (fun (sty, sty', loc) ->
-      transl_simple_type env ~closed:false Mode.Alloc.Const.legacy sty,
-      transl_simple_type env ~closed:false Mode.Alloc.Const.legacy sty', loc)
+      transl_simple_type ~new_var_jkind:Any env ~closed:false Mode.Alloc.Const.legacy sty,
```
Comment (goldfirere) [inline ocaml/typing/typedecl.ml:629]: Definitely worth a comment explaining the choices here.

Even better, put an explanation on `transl_simple_type` that walks the reader through some of our design process; a reader who is wondering why a particular choice is made will surely check for comments on `transl_simple_type`, and those comments will explain the situation.

File: ocaml/typing/typedecl.ml [hunk 3]
```diff
@@ -2355,10 +2355,10 @@ let transl_with_constraint id ?fixed_row_path ~sig_env ~sig_decl ~outer_env
   let constraints =
     List.map (fun (ty, ty', loc) ->
       let cty =
-        transl_simple_type env ~closed:false Mode.Alloc.Const.legacy ty
+        transl_simple_type ~new_var_jkind:Sort env ~closed:false Mode.Alloc.Const.legacy ty
```
Comment (goldfirere) [inline ocaml/typing/typedecl.ml:2358]: This is the LHS of a constraint; use `Any`.

File: ocaml/typing/typedecl.ml [hunk 4]
```diff
@@ -2372,7 +2372,7 @@ let transl_with_constraint id ?fixed_row_path ~sig_env ~sig_decl ~outer_env
       None -> None, None
     | Some sty ->
       let cty =
-        transl_simple_type env ~closed:no_row Mode.Alloc.Const.legacy sty
+        transl_simple_type ~new_var_jkind:Sort env ~closed:no_row Mode.Alloc.Const.legacy sty
```
Comment (goldfirere) [inline ocaml/typing/typedecl.ml:2375]: This is a manifest. Use `Any`.

File: ocaml/typing/typeclass.ml [hunk 5]
```diff
@@ -261,9 +261,9 @@ let unify_delayed_method_type loc env label ty expected_ty=
       raise(Error(loc, env, Field_type_mismatch ("method", label, trace)))
 
 let type_constraint val_env sty sty' loc =
-  let cty  = transl_simple_type val_env ~closed:false Alloc.Const.legacy sty in
+  let cty  = transl_simple_type ~new_var_jkind:Sort val_env ~closed:false Alloc.Const.legacy sty in
```
Comment (goldfirere) [inline ocaml/typing/typeclass.ml:264]: This is the LHS of a constraint. Use `Any`.

File: ocaml/typing/typeclass.ml [hunk 6]
```diff
@@ -343,7 +343,7 @@ let rec class_type_field env sign self_scope ctf =
                  ) :: !delayed_meth_specs;
                Tctf_method (lab, priv, virt, returned_cty)
            | _ ->
-               let cty = transl_simple_type env ~closed:false Alloc.Const.legacy sty in
+               let cty = transl_simple_type ~new_var_jkind:Sort env ~closed:false Alloc.Const.legacy sty in
```
Comment (goldfirere) [inline ocaml/typing/typeclass.ml:346]: This is an annotation on an expression. Use `Any`.

File: ocaml/typing/typeclass.ml [hunk 7]
```diff
@@ -367,7 +367,7 @@ and class_signature virt env pcsig self_scope loc =
   (* Introduce a dummy method preventing self type from being closed. *)
   Ctype.add_dummy_method env ~scope:self_scope sign;
 
-  let self_cty = transl_simple_type env ~closed:false Alloc.Const.legacy sty in
+  let self_cty = transl_simple_type ~new_var_jkind:Sort env ~closed:false Alloc.Const.legacy sty in
```
Comment (goldfirere) [inline ocaml/typing/typeclass.ml:370]: Use `Any` here; it will get unified very shortly.

File: ocaml/typing/typeclass.ml [hunk 8]
```diff
@@ -417,7 +417,7 @@ and class_type_aux env virt self_scope scty =
                                                    List.length styl)));
       let ctys = List.map2
         (fun sty ty ->
-          let cty' = transl_simple_type env ~closed:false Alloc.Const.legacy sty in
+          let cty' = transl_simple_type ~new_var_jkind:Sort env ~closed:false Alloc.Const.legacy sty in
```
Comment (goldfirere) [inline ocaml/typing/typeclass.ml:420]: I think (quite low confidence) that this should be `Any`. A class type can appear as an annotation on a class expression (where `Any` is appropriate) or on the RHS of a `class type` definition. In the latter case, the class type is very like a manifest: any free type variables will be in scope elsewhere.

File: ocaml/typing/typeclass.ml [hunk 9]
```diff
@@ -437,7 +437,7 @@ and class_type_aux env virt self_scope scty =
       cltyp (Tcty_signature clsig) typ
 
   | Pcty_arrow (l, sty, scty) ->
-      let cty = transl_simple_type env ~closed:false Alloc.Const.legacy sty in
+      let cty = transl_simple_type ~new_var_jkind:Sort env ~closed:false Alloc.Const.legacy sty in
```
Comment (goldfirere) [inline ocaml/typing/typeclass.ml:440]: As above, I think `Any` would be better here.

File: ocaml/typing/typeclass.ml [hunk 10]
```diff
@@ -670,7 +670,7 @@ let rec class_field_first_pass self_loc cl_num sign self_scope acc cf =
         (fun () ->
            let cty =
              Ctype.with_local_level_if_principal
-               (fun () -> Typetexp.transl_simple_type val_env
+               (fun () -> Typetexp.transl_simple_type ~new_var_jkind:Sort val_env
```
Comment (goldfirere) [inline ocaml/typing/typeclass.ml:673]: Use `Any` here: it's like an annotation on an expression.

File: ocaml/typing/typeclass.ml [hunk 11]
```diff
@@ -760,7 +760,7 @@ let rec class_field_first_pass self_loc cl_num sign self_scope acc cf =
       with_attrs
         (fun () ->
            let sty = Ast_helper.Typ.force_poly sty in
-           let cty = transl_simple_type val_env ~closed:false Alloc.Const.legacy sty in
+           let cty = transl_simple_type ~new_var_jkind:Sort val_env ~closed:false Alloc.Const.legacy sty in
```
Comment (goldfirere) [inline ocaml/typing/typeclass.ml:763]: I think this should be `Any`, similarly.

File: ocaml/typing/typeclass.ml [hunk 12]
```diff
@@ -800,7 +800,7 @@ let rec class_field_first_pass self_loc cl_num sign self_scope acc cf =
              | Some sty ->
                  let sty = Ast_helper.Typ.force_poly sty in
                  let cty' =
-                   Typetexp.transl_simple_type val_env ~closed:false Alloc.Const.legacy sty
+                   Typetexp.transl_simple_type ~new_var_jkind:Sort val_env ~closed:false Alloc.Const.legacy sty
```
Comment (goldfirere) [inline ocaml/typing/typeclass.ml:803]: I think `Any`

File: ocaml/typing/typeclass.ml [hunk 13]
```diff
@@ -1118,7 +1118,7 @@ and class_expr_aux cl_num val_env met_env virt self_scope scl =
       if Path.same decl.cty_path unbound_class then
         raise(Error(scl.pcl_loc, val_env, Unbound_class_2 lid.txt));
       let tyl = List.map
-          (fun sty -> transl_simple_type val_env ~closed:false Alloc.Const.legacy sty)
+          (fun sty -> transl_simple_type ~new_var_jkind:Sort val_env ~closed:false Alloc.Const.legacy sty)
```
Comment (goldfirere) [inline ocaml/typing/typeclass.ml:1121]: I think `Any`

**Review by goldfirere — COMMENTED** [review]
A few changes requested on choices of policy.

I did not review testsuite changes, as I wish to review them in #2107.

File: ocaml/typing/typedecl.ml [hunk 14]
```diff
@@ -360,7 +360,7 @@ let transl_labels env univars closed lbls =
     Builtin_attributes.warning_scope attrs
       (fun () ->
          let arg = Ast_helper.Typ.force_poly arg in
-         let cty = transl_simple_type env ?univars ~closed Mode.Alloc.Const.legacy arg in
+         let cty = transl_simple_type ~new_var_jkind:Sort env ?univars ~closed Mode.Alloc.Const.legacy arg in
```
[the remark this answers is not in the document]
Comment (alanechang) [reply ocaml/typing/typedecl.ml]: The claim `closed == true` iff `univars == Some _` doesn't seem to hold:

https://github.com/ocaml-flambda/flambda-backend/blob/e37e0503787855e6ffeb01266ec67aa0dc17c24c/ocaml/typing/typedecl.ml#L442

Adding a `~new_var_jkind` argument to `transl_labels`, `transl_types_gf`, `transl_constructor_arguments` and passing in either `Sort` or `Any` at the outer most call site seems to be the most straight forward approach.

Call sites:

https://github.com/ocaml-flambda/flambda-backend/blob/4a8df8efbd069bc58f32df274bc24a8053b9fd0b/ocaml/typing/typedecl.ml#L742-L752

https://github.com/ocaml-flambda/flambda-backend/blob/4a8df8efbd069bc58f32df274bc24a8053b9fd0b/ocaml/typing/typedecl.ml#L439-L444

https://github.com/ocaml-flambda/flambda-backend/blob/4a8df8efbd069bc58f32df274bc24a8053b9fd0b/ocaml/typing/typedecl.ml#L466-L469

Also added tests to check for the desired behavior.

File: ocaml/typing/typetexp.ml [hunk 15]
```diff
@@ -1270,7 +1313,7 @@ let transl_type_scheme_mono env styp =
   let typ =
     with_local_level begin fun () ->
       TyVarEnv.reset ();
-      transl_simple_type env ~closed:false Alloc.Const.legacy styp
+      transl_simple_type ~new_var_jkind:Sort env ~closed:false Alloc.Const.legacy styp
```
[the remark this answers is not in the document]
Comment (alanechang) [reply ocaml/typing/typetexp.ml:1311]: Added note to `transl_simple_type`, should there be more comments at the various call sites?

File: ocaml/typing/typedecl.ml (hunk 14 above)
[the remark this answers is not in the document]
Comment (goldfirere) [reply ocaml/typing/typedecl.ml]: Yes, you're right -- I didn't read the code carefully. Will review the code shortly.

File: ocaml/typing/typetexp.ml (hunk 15 above)
[the remark this answers is not in the document]
Comment (goldfirere) [reply ocaml/typing/typetexp.ml:1311]: Nope. No more commentary needed.

**Pull request merged**